File: src/android/app/src/main/java/com/crickettechnology/trees/Species.java
package com.crickettechnology.trees;

import org.json.*;
import android.graphics.*;

class Species 
{
    public Species(JSONObject json)
    {
        this.botanicalName = json.optString("name_botanical");
        this.botanicalFamilyName = json.optJSONObject("hierarchy").optString("family");
        JSONArray commonNames = json.optJSONArray("name_common");
        this.commonName = commonNames.length() > 0 ? commonNames.optString(0) : ""; // XXX TODO show all
        this.wikipediaTitle = json.optString("wp_link");

        JSONArray colorArray= json.optJSONArray("color");
        int r = (int) (colorArray.optDouble(0) * 255);
        int g = (int) (colorArray.optDouble(1) * 255);
        int b = (int) (colorArray.optDouble(2) * 255);
        this.markerColor = Color.rgb(r, g, b);
    }

    public final String botanicalName;
    public final String botanicalFamilyName;
    public final String commonName;
    public final String wikipediaTitle;
    public final int markerColor;
}

File: src/ios/trees/Species.swift
import Foundation
import UIKit

class Species: Equatable
{
    init(json: JsonObject)
    {
        self.botanicalName = json["name_botanical"] as! String
        self.commonNames = json["name_common"] as! [String]
        self.wikipediaTitle = json["wp_link"] as! String
        self.treeCount = json["count"] as! Int

        let color = (json["color"] as! [NSNumber]).map({ CGFloat($0.floatValue) })
        self.markerColor = UIColor(red:color[0], green:color[1], blue:color[2], alpha:1.0)

        let hierJson = json["hierarchy"] as! JsonObject
        var hier = [Taxon]()
        func add(rank: TaxonomicRank)
        {
            if let name = hierJson[rank.rawValue] as? String
            {
                let wpTitle = hierJson[rank.rawValue + "_wp"] as? String
                hier.append(Taxon(rank:rank, name:name, wikipediaTitle:wpTitle))
            }
        }
        for rank in TaxonomicRank.allCases
        {
            add(rank: rank)
        }
        self.hierarchy = hier
    }

    public let botanicalName: String
    public let commonNames: [String]
    public let wikipediaTitle: String
    public let markerColor: UIColor
    public let treeCount: Int

    public let hierarchy: [Taxon]


    /*
    ////////////////////////////////////////
    // Hashable

    var hashValue: Int 
    { 
        get { return self.botanicalName.hashValue }
    }
    */


    ////////////////////////////////////////
    // Equatable

    static func == (lhs: Species, rhs: Species) -> Bool
    {
        return lhs.botanicalName == rhs.botanicalName
    }

}
